pennylane/_qubit_device.py

```diff
@@ -29,100 +30,164 @@
+        implementing a suitable subset of
+
+        * :meth:`apply`
 
-        For plugin developers: Instead of overwriting this, consider implementing a suitable subset of
-        :meth:`pre_apply`, :meth:`apply`, :meth:`post_apply`, :meth:`pre_measure`,
-        :meth:`expval`, :meth:`var`, :meth:`sample`, :meth:`post_measure`, and :meth:`execution_context`.
+        * :meth:`~.generate_samples`
+
+        * :meth:`~.probability`
 
         Args:
-            queue (Iterable[~.operation.Operation]): operations to execute on the device
-            observables (Iterable[~.operation.Observable]): observables to measure and return
-            parameters (dict[int->list[ParameterDependency]]): Mapping from free parameter index to the list of
-                :class:`Operations <pennylane.operation.Operation>` (in the queue) that depend on it.
+            circuit (~.CircuitGraph): circuit to execute on the device
 
         Raises:
             QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
 
         Returns:
             array[float]: measured value(s)
         """
-        if parameters is None:
-            parameters = {}
-
-        self.check_validity(queue, observables)
-        self._op_queue = queue
-        self._obs_queue = observables
-        self._parameters = {}
-        self._parameters.update(parameters)
+        self.check_validity(circuit.operations, circuit.observables)
 
         with self.execution_context():
-            self.pre_apply()
-
-            for operation in queue:
-                # Pass instances directly
-                self.apply(operation)
-
-            self.rotate_basis(observables)
-
-            self.post_apply()
+            # determine operations required to diagonalize observables
+            rotations = self.rotate_basis(circuit.observables)
 
-            self.pre_measure()
+            # apply all circuit operations
+            self.apply(circuit.operations, rotations)
 
-            results = self.statistics(observables)
+            # generate computational basis samples
+            self.generate_samples()
 
-            self.post_measure()
-
-            self._op_queue = None
-            self._obs_queue = None
-            self._parameters = None
+            # compute the required statistics
+            results = self.statistics(circuit.observables)
 
             # Ensures that a combination with sample does not put
             # expvals and vars in superfluous arrays
-            sample_return_types = (obs.return_type is Sample for obs in observables)
+            sample_return_types = (obs.return_type is Sample for obs in circuit.observables)
             if any(sample_return_types) and not all(sample_return_types):
                 return self._asarray(results, dtype="object")
 
             return self._asarray(results)
 
+    @abc.abstractmethod
+    def apply(self, operations, rotations=None, **kwargs):
```

Comment:
It just came to my mind, based on the current structure of ``execute``, what would be the reason for having a separate ``rotations`` argument here? These ``rotations`` would not be treated differently than other operations inside of the circuit (except for the fact that it would be sure that state preparations are not among them, however, we would like to treat state preparations differently anyways in the long run).

Reply:
The only reason is to give the plugin the choice of whether to implement the rotations or not :slightly_smiling_face: Some plugins have their own expectation value computation which may be more efficient that the inherited rotation based method, or may support measurements outside of the computational basis.

Reply:
Just came to my mind: it might be worth adding this information in the docstring (i.e. to explain the ``rotations=None`` case)